### PR TITLE
Issue #4197: Allow per content type search indexing.

### LINKED ISSUE #4197: [UX] Add configuration to exclude node types from search indexing
**Description of the need**

Here is the excerpt from https://github.com/backdrop-contrib/search_index_options:

> The core search module scans all the available content, regardless of whether some content types were meant to be searchable at all or not. This kind of design causes extra processing load while indexing and storing lots of unnecessary information in database. 

**Proposed solution**

The quote goes on as:

> The Search Index Options module addresses this problem by letting users to set only desired content types for indexing, thus significantly reducing the load and amount of stored index data.

I suggest to save Backdrop from repeating one of the most illogical features Drupal has - it takes all the nodes into database, and only after that the modules start interacting and operating with the search index data. Why can't we let users to choose which content types or nodes they want to be indexed?! If you look at the module's code it's pretty simple and straightforward. So I suggest to move some features of the module to core. 

**Additional information**

Here is the stats screenshot of a test Backdrop site with 52 nodes in total, 24 of which are of **Page** type (indexed) and rest of **Post** type (not indexed). 

Screen Shot 2019-11-06 at 1 42 36 AM

With Post content type not indexed at all, we store only 46% of initial data in the database. Remember those huge index tables of Drupal sites? Now, we can do it right way on Backdrop core!

---

~PR by @alanmels https://github.com/backdrop/backdrop/pull/2990~
PR by @quicksketch https://github.com/backdrop/backdrop/pull/3129 (based on @alanmels's PR)

### PR DESCRIPTION
**Pull request by quicksketch**
Fixes https://github.com/backdrop/backdrop-issues/issues/4197
Replaces https://github.com/backdrop/backdrop/pull/2990

### CODE REVIEW

File: core/modules/node/node.theme.inc [hunk 1]
```diff
@@ -37,27 +37,27 @@ function theme_node_preview_banner_form($variables) {
...
-    $row[] = $form['factors'][$key]['#title'];
-    $form['factors'][$key]['#title_display'] = 'invisible';
-    $row[] = backdrop_render($form['factors'][$key]);
+    $row[] = $element[$key]['#title'];
+    $element[$key]['#title_display'] = 'invisible';
+    $row[] = backdrop_render($element[$key]);
     $rows[] = $row;
   }
+  $output = '';
+  $output .= theme('help', array('markup' => $element['#help']));
```
Comment (jenlampton) [inline core/modules/node/node.theme.inc:58]: love it!

File: core/modules/search/search.admin.inc [hunk 2]
```diff
@@ -77,8 +76,9 @@ function search_admin_settings($form, &$form_state) {
     '#type' => 'radios',
     '#default_value' => $config->get('search_default_module'),
     '#options' => $module_options,
-    '#description' => t('Only one type of item will be searched by default. This selection will determine what appears at http://www.example.com/search.')
+    '#description' => t('Only one type of item will be searched by default. This selection will determine what appears at <a href="@search">@search</a>.', array('@search' => url('search', array('absolute' => TRUE)))),
```
Comment (jenlampton) [inline core/modules/search/search.admin.inc:79]: nice improvement here :)

File: core/modules/search/search.module [hunk 3]
```diff
@@ -412,13 +412,14 @@ function search_dirty($word = NULL) {
...
-  // We register a shutdown function to ensure that search_total is always up
-  // to date.
+  // We register a shutdown function to ensure that search_total is always up-to-date.
   backdrop_register_shutdown_function('search_update_totals');
 
-  foreach (config_get('search.settings', 'search_active_modules') as $module) {
-    // Update word index
-    module_invoke($module, 'update_index');
+  $search_active_modules = config_get('search.settings', 'search_active_modules');
+  if (isset($search_active_modules)) {
```
Comment (jenlampton) [inline core/modules/search/search.module:419]: nice microptimization here.

**Comment by quicksketch** [comment]
Merged into 1.x for 1.16.0.

**Pull request merged**